### DISCUSSION TITLE
tesr to mesh

### DISCUSSION THREAD

**Discussion by majidkavousi** [neperfepx/neper#702] [Q&A] [answered]
Hi Romain,

I have a general question which is not directly related to neper. But I would be really thankful if you could help me with that. 
Below is a picture of a microstructure from a manually written tesr file and then meshed with neper with two element sizes. I'm wondering, what is the process or algorithm for creating a mesh from a voxel/pixel-based data file. I assume the boundaries of grains are approximated by some polygons at first. It would be great if you could give me some insights about this process in general.

![image](https://github.com/neperfepx/neper/assets/99205979/f29e785e-60be-4eb6-8758-d25566653913)

**Comment by rquey** [comment, marked as answer]
I'm glad you are showing these pictures.  In 2D, a tesr can indeed be meshed directly. It is not the same process as in 3D, where the tesr must be approximated by a tesr beforehand.

In 2D, Neper determines the topology from the tesr (i.e., where the grain boundaries and the triple points are), and then does a standard meshing, except that the 1D meshes (of the grain boundaries)  are projected onto to actual boundaries and smoothed (Laplacian smoothing).
